src/rust/engine/fs/store/src/snapshot.rs

```diff
@@ -559,6 +560,73 @@ impl Snapshot {
       })
       .to_boxed()
   }
+
+  pub fn get_snapshot_subset(
+    store: Store,
+    digest: Digest,
+    path_globs: PathGlobs,
+    workunit_store: WorkUnitStore,
+  ) -> BoxFuture<Snapshot, String> {
+    use bazel_protos::remote_execution::{Directory, DirectoryNode, FileNode};
+
+    let traverser = move |_: &Store,
+                          path_so_far: &PathBuf,
+                          _: Digest,
+                          directory: &Directory|
+          -> BoxFuture<(Vec<PathStat>, StoreManyFileDigests), String> {
+
+      let subdir_paths: Vec<PathBuf> = directory
+        .get_directories()
+        .iter()
+        .map(move |node: &DirectoryNode| path_so_far.join(node.get_name()))
+        .filter(|path: &PathBuf| path_globs.matches(&[path.clone()]))
+        .collect();
+
+      let file_paths: Vec<(PathBuf, Result<Digest, String>, bool)> = directory
+        .get_files()
+        .iter()
+        .map(|node: &FileNode| (path_so_far.join(node.get_name()), node.get_digest().into(), node.is_executable))
+        .filter(|(path, _, _)| path_globs.matches(&[path.clone()]))
+        .collect();
+
+      let mut path_stats: Vec<PathStat> = vec![];
+      for path in subdir_paths.into_iter() {
+        path_stats.push(PathStat::dir(path.clone(), Dir(path)));
+      }
+
+      let mut hash = HashMap::new();
+      for (path, maybe_digest, is_executable) in file_paths.into_iter() {
+        let digest = match maybe_digest {
+          Ok(d) => d,
+          Err(err) => return future::err(err).to_boxed()
+        };
+        hash.insert(path.clone(), digest);
```

Comment:
You could probably have walk return `Vec<Vec<(PathStat, Digest)>>`, and then convert it to a hashmap only at the end? Not a big deal.

Reply:
I think there's some optimization that could be done here, but I don't want to spend too much more time hyper-optimizing this commit.